frontend/app/javascript/controllers/coupon_code_controller.js

```diff
@@ -0,0 +1,67 @@
+import { Controller } from 'stimulus'
+
+export default class extends Controller {
+  static targets = [ 'code', 'resultText', 'resultIcon' ]
+
+  get code() {
+    return this.codeTarget.value.trim()
+  }
+
+  set code(value) {
+    this.codeTarget.value = value
+  }
+
+  initialize() {
+    this.apiClient = window.SpreeSDK.makeClient({
+      host: window.SpreeAPI.storefrontHost
+    })
+  }
+
+  async apply() {
+    if (!this.code) return
+
+    const orderToken = window.SpreeAPI.orderToken
+
+    const response = await this.apiClient.cart.applyCouponCode({ orderToken }, {
+      coupon_code: this.code
+    })
+
+    this.handleResponse(response)
+  }
+
+  async remove() {
+    if (!this.code) return
+
+    const orderToken = window.SpreeAPI.orderToken
+
+    const response = await this.apiClient.cart.removeCouponCode({ orderToken }, this.code)
+
+    this.handleResponse(response)
+  }
+
+  handleResponse(response) {
+    if (response.isFail()) {
+      this.handleError(response.fail().serverResponse.data.error)
```

Comment:
To handle all Spree responses, including unexpected ones, a check like `response.fail() instanceof window.SpreeSDK.errors.BasicSpreeError` is needed here. If `response.fail()` is an instance of `BasicSpreeError` or its subclass, then `response.fail().summary` is the same as `response.fail().serverResponse.data.error`. If `response.fail()` is a superclass of `BasicSpreeError` it means there was a more serious error.